Update phone validation regex to accept only numbers and correct amount - Fix "must consist of only numbers" validation error message to be specific

MaxwellGarceau opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

## Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->
Update phone validation to fix several bugs and give detailed error messages.

#### Issue #101 - Update phone validation regex to accept only numbers and correct amount
- Fix a bug that would allow a user to enter phone numbers that contained letters.

#### Issue #104  - Update phone validation regex to accept only numbers and correct amount
- Fix a bug that would allow a user to enter whitespace in the middle of the last group of numbers
- If too few digits are entered, the form displays an appropriate error message letting the user know that too few digits were entered

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #101 
Closes #104  
_Both these issues required work in the same function that would have created ugly merge conflicts._

## How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->
### Setup
1. Change the phone format in the test user Mailchimp account to US format
2. Click the "Update List" button in the Mailchimp admin
3. Set the Phone number merge field to be included
4. Navigate to a FE form for submission testing

### General
- Form submission with empty phone number and only email should succeed

### Phone number should not accept letters

#### Phone number should not accept letters - Bug is fixed
- Submit a form with a phone number that contains a letter. **The submission should fail.** The error message should notify the user that the phone number must only contain numbers.

#### Test a valid phone number
- Submit a valid US phone number and the submission should succeed

#### Test that other invalid characters are not accepted
The phone number field should only accept numbers. The error message should notify the user that the phone number must only contain numbers.
- The phone field should not accept any of these characters
  	•	Letters (A–Z, a–z).
	•	Special characters like @, #, $, %, *, &, !, etc.
	•	Whitespace, both at the ends and also in the middle.
	•	Multiple or misplaced delimiters, such as --, (123)), or leading/trailing symbols.
	•	Any character outside the ASCII range (e.g., emojis, non-standard symbols). - **too broad to really test comprehensively**

### Phone number does not contain enough digits

#### Whitespace testing
All whitespace should be stripped and should return an error notifying the user that the phone number does not have enough digits.
- `1 3` - `12 ` - ` 234` should fail validation with an **error messaging stating that there are not enough digits**

## Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Added - New feature
> Changed - Existing functionality
> Deprecated - Soon-to-be removed feature
> Removed - Feature
> Fixed - Bug fix
> Security - Vulnerability

## Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @username, @username2, ...

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/mailchimp/wordpress/blob/develop/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

MaxwellGarceau commented:
@dkotter I think this ticket would be a good candidate for some PHP unit tests around the address validation. The changes are small and the only WP dependency is WP_Error. What do you think?

Assuming that you think it's worth adding the tests, I wanted to ask if 10up had a preferred testing library or "10up way" of unit testing? There's currently no PHP tests in this project.

My preference would be to use Pest with WP_Mock or Brain Monkey, but I'm open to anything.

dkotter commented on mailchimp.php:
```diff
@@ -1046,26 +1046,55 @@ function mailchimp_sf_merge_submit( $mv ) {
 
+	// Trim whitespace
+	$opt_val = array_map( 'trim', $opt_val ); // Beginning and end
+	$opt_val = array_map( fn( $s ) => preg_replace( '/\s/', '', $s ), $opt_val ); // Middle
```
Our minimum PHP version is still 7.0 and I believe arrow functions weren't introduced until 7.4, so we'll need to change this to use an anonymous function I believe

MaxwellGarceau replied:
Nice catch! I've updated the function.

MaxwellGarceau commented:
## Arrow functions not caught by PHPCS
**TLDR:**
- PHPCompatibilityWP isn't catching arrow functions locally or in the CI/CD pipeline.
- Tried various fixes and nothing worked.
- One workaround is using rector in `--dry-run` mode.
- Should I submit a bug report in the [PHPCompatibility repo](https://github.com/PHPCompatibility/PHPCompatibility) or do you think it's more likely some kind of user/config error on my end?

### Verifying the problem
@dkotter while updating the arrow function I noticed that the current PHPCS compatibility check doesn't catch arrow functions. The specific CLI command I was running was `phpcs --standard=./phpcs-compat.xml -p -s .`. I tried this command both outside the docker container on my local machine running PHP 8.4 and inside the wp-env docker container.

I also tried PHP 5.0 as the test version to ensure the XML file was working and the linter picked up on several incompatibilities.

### Solution attempts
I tried to explicitly add the arrow function rule, tried other rulesets (PHPCompatibility), and tried adding a custom sniff to register the `T_FN` keyword for the token value PHPCS looks for. I couldn't get the arrow functions to register as errors on PHP 7.0.

#### PRs in PHPCompatibility adding arrow function support
I looked through some of the merged PRs in the PHPCompatibility codebase and it seems like arrow functions should trigger on the `fn` keyword.
- https://github.com/PHPCompatibility/PHPCompatibility/pull/1593/
- https://github.com/squizlabs/PHP_CodeSniffer/issues/2523

## Workaround: Rector
I did a quick test with rector and rector catches arrow functions. An alternative would be using rector in `--dry-run` mode and failing the Github action if any diffs came back.

Not the best solution and kind of a hack, but may be nice to have for easy PHP upgrades in the future as well.

<img width="1029" alt="Screenshot 2025-01-07 at 3 24 03 PM" src="https://github.com/user-attachments/assets/e2148071-a024-4fba-86cf-066f0ff5a4e3" />

#### Bug report for PHPCompatibility
Do you think this is worth making a bug report in the [PHPCompatibility codebase](https://github.com/PHPCompatibility/PHPCompatibility)? It's definitely possible I'm just missing something obvious.

dkotter commented:
> Do you think this is worth making a bug report in the [PHPCompatibility codebase](https://github.com/PHPCompatibility/PHPCompatibility)? It's definitely possible I'm just missing something obvious.

I'm pretty sure the issue here is the PHPCompat repo hasn't pushed out a new release since December 2019, so all new sniffs added since then don't actually run if you're not using a development version of the tool (which is unfortunate). The easiest way to address this (which I should have done when I was setting up PHPCS on this repo) is to install the dev version of that tool.

Can see instructions on how to do that here: https://github.com/10up/phpcs-composer/issues/52. Basically just need to run:
`composer require --dev phpcompatibility/php-compatibility:"dev-develop as 9.99.99"`

MaxwellGarceau commented on includes/validation/functions.php:
```diff
@@ -0,0 +1,41 @@
+<?php
```
Kept the original functions here for backwards compatibility. The idea is work towards more class based modularization for easier unit testing.

After the composer autoload changes go into develop it would be great to get a DI container involved to avoid direct class instantiation in the function.

MaxwellGarceau commented on tests/unit/USPhoneNumberValidationTest.php:
```diff
@@ -0,0 +1,176 @@
+		$wp_error = Mockery::mock('WP_Error');
+
+		// Step 2: Mock the factory that creates WP_Error.
+		$wp_error_factory = function ( $code, $message, $data = null ) use ( $wp_error ) {
+			// Step 3: Make assertions against the properties we want the WP_Error object to contain
+			// and dynamically fill them to return what our validate functions will pass in.
+			$wp_error
+				->shouldReceive('get_error_code')
+				->andReturn($code);
+			$wp_error
+				->shouldReceive('get_error_message')
+				->with($code)
+				->andReturn($message);
+			return $wp_error;
+		};
```
We're dynamically filling the WP_Error assertions here with what we actually get back from the validate functions.

MaxwellGarceau commented on tests/unit/USPhoneNumberValidationTest.php:
```diff
@@ -0,0 +1,176 @@
+		// Is WP_Error
+		$this->assertInstanceOf(WP_Error::class, $result, "Result (Not WP_Error): {$result}");
+
+		// Error code
+		$this->assertEquals(Validate_Merge_Fields::PHONE_VALIDATION_ERROR_CODE, $result->get_error_code());
+
+		// Error message
+		$this->assertMatchesRegularExpression(
+			'/must consist of only numbers/',
+			$result->get_error_message(Validate_Merge_Fields::PHONE_VALIDATION_ERROR_CODE)
+		);
```
Here we're testing the dynamically filled WP_Error mock with what we really should be getting back from the validation functions.

MaxwellGarceau commented on tests/unit/bootstrap.php:
```diff
@@ -0,0 +1,50 @@
+// (There are more)
+
+/**
+ * Strategy:
+ * - Unit testing only (no integration or end-to-end).
+ * - Mock WP core functions with WP_Mock (WP core is not loaded).
+ * - Promote long-term maintainability through modularization.
+ * 
+ * Reasoning:
+ * This plugin is not modular and initializing the plugin
+ * will require mocking all of the WP core functions
+ * used in order to avoid fatal errors on loading.
+ * 
+ * Instead, we'll require functionality as it's needed
+ * by first:
+ * 
+ * 1) Modularizing the functionality into a namespaced file and then
+ * 2) Importing it directly into the test.
+ * 
+ * End to end tests are handled in Cypress. Integration
+ * and E2E tests should be handled in Cypress.
```
I wrote a little strategy here for long-term app health that could be moved to over time. What do you think?

MaxwellGarceau reviewed (commented):
@dkotter I've added PHPUnit, WP_Mock, and created the tests for the validation logic. I moved a few functions out of `mailchimp.php` into their own namespaced files and turned them into classes for easier testing.

I highlighted the most import parts. It's kind of a large PR and I'm happy to make any changes you think would be beneficial.

dkotter commented on mailchimp.php:
```diff
@@ -1041,65 +1053,6 @@ function mailchimp_sf_merge_submit( $mv ) {
- * @param array $data Data
- * @return void
- */
-function mailchimp_sf_merge_validate_phone( $opt_val, $data ) {
```
For any of these functions that we're removing/moving, technically someone else could be using these functions themselves and this would cause a fatal error for them. What are your thoughts on deprecating these instead of fully removing?

MaxwellGarceau replied:
Great idea! I've added back the original functions with the deprecated comments. In addition, I've moved them into `/includes/deprecated/deprecated-functions.php` so that we can isolate them for easier future removal.

dkotter commented on includes/deprecated/deprecated-functions.php:
```diff
@@ -0,0 +1,55 @@
+ * @return void|WP_Error Returns an error object if validation fails, otherwise processes the value.
+ */
+function mailchimp_sf_merge_validate_phone( $opt_val, $data ) {
+	$validator = new Validate_Merge_Fields();
```
For any function we deprecate, we should run [`_deprecated_function` ](https://developer.wordpress.org/reference/functions/_deprecated_function/) for those so users are made aware of the deprecation

MaxwellGarceau replied:
Nice catch. I've added the `_deprecated_function` calls to `mailchimp_sf_merge_validate_phone`, `mailchimp_sf_merge_validate_address`, and `mailchimp_sf_merge_validate_address`.

dkotter commented on includes/deprecated/deprecated-functions.php:
```diff
@@ -0,0 +1,55 @@
+function mailchimp_sf_merge_validate_address( $opt_val, $data ) {
+	$validator = new Validate_Merge_Fields();
+	return $validator->validate_address( $opt_val, $data );
+}
```
We probably want to deprecate the `mailchimp_sf_where_am_i` function as well

MaxwellGarceau replied:
`mailchimp_sf_merge_validate_address` is now in the `includes/deprecated/deprecated-functions.php` file with a `_deprecated_function` call and docblock.

For a quick test, I re-added `mailchimp_sf_where_am_i` back to `mailchimp.php` where it had originally been and the plugin is loading along with the new deprecation message.

<img width="1710" alt="Screenshot 2025-01-14 at 10 51 49 PM" src="https://github.com/user-attachments/assets/1b6707a3-89c4-46b4-9c6d-123262bca2ab" />

![Screenshot 2025-01-14 at 10 51 49 PM (2)](https://github.com/user-attachments/assets/28690ffa-7703-4bad-a7d5-b3afeb7b54cb)

qasumitbagthariya commented:
QA Update: ✅
---

I have verified this PR in the `bug/issue-101-and-104/fix-phone-number-validation-and-error-messages` branch which has been fixed and is functioning as intended.


- **Form Submission Without Phone Number:** ✅

 
<img width="1024" alt="image" src="https://github.com/user-attachments/assets/71581551-d4c9-41c7-9af9-6455a51494a2" />

- Phone Number Accepts Only Valid Inputs:  ✅

- Letters: e.g., "ABC123".
- Special Characters: e.g., "@123$".
- Whitespace: e.g., " 123 456 ".
- Invalid Delimiters: e.g., "12--34", "(123))", or "123-456-".


<img width="879" alt="image" src="https://github.com/user-attachments/assets/3133dea9-a652-4383-b0e8-fa08df7f0f52" />

Accepts only numeric characters (0–9).  ✅

Successfully accepts valid US-format phone numbers. ✅

**Testing Environment** 

<Details>

- WordPress: 6.7.1
- Theme: Twenty Twenty-Four 1.3
- WooCommerce - 9.5.1
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: bug/issue-101-and-104/fix-phone-number-validation-and-error-messages

</Details>

MaxwellGarceau commented:
> @MaxwellGarceau thanks for the PR! Could you please rebase your PR on top of the latest changes in the base branch?

Rebased `bug/issue-101-and-104/fix-phone-number-validation-and-error-messages` with changes from `develop`.

Used rebase specifically because I was prompted by the PR bot, but let me know if another method is preferred in the future. A backup of the original branch can be found at [before-rebase/bug/issue-101-and-104/fix-phone-number-validation-and-error-messages](https://github.com/mailchimp/wordpress/tree/before-rebase/bug/issue-101-and-104/fix-phone-number-validation-and-error-messages).

Fixed merge conflicts and verified changes by testing validation and submission locally with the GUI as well as running the PHP Unit phone validation tests.

qasumitbagthariya commented:
@dkotter @MaxwellGarceau 
Is another round of testing required?

dkotter commented:
> @dkotter Is another round of testing required?

@qasumitbagthariya Not at the moment, no. Holding this in UAT for now

dkotter commented:
> @dkotter do we need to include this in the 1.7.0 release? If so, can we merge this to smoke-testing branch for regression testing?

No, purposely holding this in UAT for now

jeffpaul commented:
Closing this PR as there's a bit simpler, more straightforward approach that could be taken that doesn't involve as much refactoring and moving of things within the plugin.  Thanks for the work here Max, but I'm going to move the associated issues to the backlog for us to consider an alternate approach as time allows.
